### PR TITLE
*: deal with the noise caused by some statistics

### PR DESCRIPTION
**Pull request by nolouch**
that caused by the following case:
1. leader changed, new leader reports the flow includes all record when it as a follower.
2. store heartbeat will report immediately when sending a snapshot
3. new split region will report all flow that record by its parent.

### CODE REVIEW

File: server/schedule/hot_cache.go [hunk 1]
```diff
@@ -134,7 +135,12 @@ func (w *HotSpotCache) isNeedUpdateStatCache(region *core.RegionInfo, flowBytes
 
 	if oldItem != nil {
 		newItem.HotDegree = oldItem.HotDegree + 1
+		// denosing
+		if newItem.StoreID != oldItem.StoreID && math.Floor(math.Log10(float64(oldItem.FlowBytes))) != math.Floor(math.Log10(float64(newItem.FlowBytes))) {
```
Comment (disksing) [inline server/schedule/hot_cache.go]: I think we need some comments about how it works.

File: server/schedule/hot_cache.go [hunk 2]
```diff
@@ -131,10 +132,19 @@ func (w *HotSpotCache) isNeedUpdateStatCache(region *core.RegionInfo, flowBytes
...
 	}
+	// notSameMagnitude to judge that two numbers are not in the same order of magnitude
+	notSameMagnitude := func(a, b float64) bool {
+		return math.Abs(math.Log10(a)-math.Log10(b)) >= 1
+	}
 
 	if oldItem != nil {
 		newItem.HotDegree = oldItem.HotDegree + 1
+		// denosing
+		if newItem.StoreID != oldItem.StoreID && notSameMagnitude(float64(oldItem.FlowBytes), float64(newItem.FlowBytes)) {
```
Comment (disksing) [inline server/schedule/hot_cache.go]: I don't think it is a good idea.
For example, if a region's flowBytes are reported as '1, 10, 10, 10 ...', you will always use 1, which is apparently not correct.

Comment (disksing) [reply server/schedule/hot_cache.go]: Maybe a simple median filter is better. See https://en.wikipedia.org/wiki/Median_filter

Comment (nolouch) [reply server/schedule/hot_cache.go]: this condition's precondition is leader change, that means the store id is not same. the normal report not yet been denosing now.

File: server/core/statistics.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"github.com/montanaflynn/stats"
+)
+
+// RollingStats is a rolling statistics.
```
Comment (disksing) [inline server/core/statistics.go]: Better to add some comments to explain what is it used for and how it works.

File: server/cache.go [hunk 4]
```diff
@@ -399,7 +409,7 @@ func (c *clusterInfo) handleRegionHeartbeat(region *core.RegionInfo) error {
 	var saveKV, saveCache, isNew bool
 	if origin == nil {
 		log.Infof("[region %d] Insert new region {%v}", region.GetId(), region)
-		saveKV, saveCache, isNew = true, true, true
+		saveKV, saveCache, isNew, isWriteUpdate, isReadUpdate = true, true, true, false, false
```
Comment (disksing) [inline server/cache.go]: Why do we need to set them to false?

File: server/cache.go [hunk 5]
```diff
@@ -369,8 +371,16 @@ func (c *clusterInfo) handleStoreHeartbeat(stats *pdpb.StoreStats) error {
 	if store == nil {
 		return errors.Trace(core.ErrStoreNotFound(storeID))
 	}
+	lastStats := store.Stats
 	store.Stats = proto.Clone(stats).(*pdpb.StoreStats)
 	store.LastHeartbeatTS = time.Now()
+	// denosing
```
Comment (disksing) [inline server/cache.go]: The comment should be more specific.

File: server/cache.go [hunk 6]
```diff
@@ -369,8 +371,16 @@ func (c *clusterInfo) handleStoreHeartbeat(stats *pdpb.StoreStats) error {
...
+	if store.Stats.GetInterval().GetEndTimestamp()-store.Stats.GetInterval().GetStartTimestamp() < minStoreHeartbeatInterval {
+		store.Stats.BytesRead = lastStats.BytesRead
+		store.Stats.BytesWritten = lastStats.BytesWritten
+		store.Stats.KeysRead = lastStats.KeysRead
+		store.Stats.KeysWritten = lastStats.KeysWritten
```
Comment (disksing) [inline server/cache.go]: Do we need to transform the store's flow bytes to bytes per second using the interval reported by tikv? Do we need to use the _RollingStatistics_ like region flow bytes?

File: server/cache.go [hunk 7]
```diff
@@ -359,6 +371,8 @@ func (c *clusterInfo) isPrepared() bool {
 	return float64(c.Regions.Length())*collectFactor <= float64(c.activeRegions)
 }
 
+const minStoreHeartbeatInterval = 1
```
Comment (disksing) [inline server/cache.go]: unused?

File: server/core/region.go [hunk 8]
```diff
@@ -323,6 +323,8 @@ type RegionStat struct {
 	AntiCount int
 	// Version used to check the region split times
 	Version uint64
+	// Stats maintain a windows's statistics
```
Comment (disksing) [inline server/core/region.go]: The comment is too ambiguous to understand.

File: server/core/statistics.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+package core
+
+import (
+	"github.com/montanaflynn/stats"
+)
+
+// RollingStats provides rolling statistics with specified windows size.
+// There have windows size records used for calculating. it provides Rolling
+// `Median` now, which used to filter noise in some situation.
```
Comment (disksing) [inline server/core/statistics.go]: Hard for me to understand... @queenypingcap Could you help to make the comment more fluent?

File: server/core/store.go [hunk 10]
```diff
@@ -424,42 +429,122 @@ func (s *StoresInfo) SetRegionSize(storeID uint64, regionSize int64) {
 	}
 }
 
-// TotalWrittenBytes return the total written bytes of all StoreInfo
-func (s *StoresInfo) TotalWrittenBytes() uint64 {
-	var totalWrittenBytes uint64
+// TotalBytesWrittenRate return the total written bytes rate of all StoreInfo.
```
Comment (disksing) [inline server/core/store.go]: return -> returns.

File: server/core/store.go [hunk 11]
```diff
@@ -424,42 +429,122 @@ func (s *StoresInfo) SetRegionSize(storeID uint64, regionSize int64) {
...
+			totalWrittenBytes += s.RollingStoreStats.GetBytesWriteRate()
 		}
 	}
 	return totalWrittenBytes
 }
 
-// TotalReadBytes return the total read bytes of all StoreInfo
-func (s *StoresInfo) TotalReadBytes() uint64 {
-	var totalReadBytes uint64
+// TotalBytesReadRate return the total read bytes rate of all StoreInfo.
```
Comment (disksing) [inline server/core/store.go]: returns.

File: server/core/store.go [hunk 12]
```diff
@@ -424,42 +429,122 @@ func (s *StoresInfo) SetRegionSize(storeID uint64, regionSize int64) {
...
 	for _, s := range s.stores {
-		res[s.GetId()] = s.Stats.GetBytesRead()
+		res[s.GetId()] = uint64(s.RollingStoreStats.GetBytesReadRate())
 	}
 	return res
 }
+
+// GetStoresKeysWriteStat return the keys write stat of all StoreInfo.
+func (s *StoresInfo) GetStoresKeysWriteStat() map[uint64]uint64 {
+	res := make(map[uint64]uint64)
```
Comment (disksing) [inline server/core/store.go]: alloc with capacity?

File: server/core/store.go [hunk 13]
```diff
@@ -424,42 +429,122 @@ func (s *StoresInfo) SetRegionSize(storeID uint64, regionSize int64) {
...
+	for _, s := range s.stores {
+		res[s.GetId()] = uint64(s.RollingStoreStats.GetKeysWriteRate())
+	}
+	return res
+}
+
+// GetStoresKeysReadStat return the bytes read stat of all StoreInfo.
+func (s *StoresInfo) GetStoresKeysReadStat() map[uint64]uint64 {
+	res := make(map[uint64]uint64)
```
Comment (disksing) [inline server/core/store.go]: alloc with capacity?

File: server/core/store.go [hunk 14]
```diff
@@ -424,42 +429,122 @@ func (s *StoresInfo) SetRegionSize(storeID uint64, regionSize int64) {
...
+func (s *StoresInfo) GetStoresBytesReadStat() map[uint64]uint64 {
 	res := make(map[uint64]uint64)
 	for _, s := range s.stores {
-		res[s.GetId()] = s.Stats.GetBytesRead()
+		res[s.GetId()] = uint64(s.RollingStoreStats.GetBytesReadRate())
 	}
 	return res
 }
+
+// GetStoresKeysWriteStat return the keys write stat of all StoreInfo.
```
Comment (disksing) [inline server/core/store.go]: returns

File: server/core/store.go [hunk 15]
```diff
@@ -424,42 +429,122 @@ func (s *StoresInfo) SetRegionSize(storeID uint64, regionSize int64) {
...
+func (s *StoresInfo) GetStoresKeysWriteStat() map[uint64]uint64 {
+	res := make(map[uint64]uint64)
+	for _, s := range s.stores {
+		res[s.GetId()] = uint64(s.RollingStoreStats.GetKeysWriteRate())
+	}
+	return res
+}
+
+// GetStoresKeysReadStat return the bytes read stat of all StoreInfo.
```
Comment (disksing) [inline server/core/store.go]: returns

**Comment by iamxy** [comment]
/run-integration_common_test tidb=release-2.0 tikv=release-2.0 tidb-test=pr/535

**Comment by iamxy** [comment]
/run-integration-common-test tidb=release-2.0 tikv=release-2.0 tidb-test=pr/535

**Comment by iamxy** [comment]
/run-integration-common-test tidb=release-2.0 tikv=release-2.0 tidb-test=pr/535

File: server/core/statistics.go (hunk 9 above)
Comment (queenypingcap) [reply server/core/statistics.go]: How about the following?
RollingStats provides rolling statistics with specified window size. There are window size records for calculating. It provides median of the statistics to filter noises in some cases.

File: server/core/statistics.go [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+func (r *RollingStats) Add(n float64) {
+	r.records[r.count%r.size] = n
+	r.count++
+}
+
+// Median returns the median of the records.
+// it can be used to filter noise.
+// References: https://en.wikipedia.org/wiki/Median_filter.
+func (r *RollingStats) Median() float64 {
+	if len(r.records) == 0 {
```
Comment (disksing) [inline server/core/statistics.go]: `r.count == 0`? `len(r.records)` will not be 0.

File: server/schedule/hot_cache.go [hunk 17]
```diff
@@ -134,26 +137,32 @@ func (w *HotSpotCache) isNeedUpdateStatCache(region *core.RegionInfo, flowBytes
...
 	}
 	if flowBytes >= hotRegionThreshold {
 		if oldItem == nil {
 			w.incMetrics("add_item", kind)
+			newItem.Stats = core.NewRollingStats(rollingWindowsSize)
 		}
-		return true, newItem
+		newItem.Stats.Add(float64(flowBytes))
+		needUpdate = true
+		return needUpdate, newItem
```
Comment (disksing) [inline server/schedule/hot_cache.go]: Why introduce the `needUpdate`? I think original `regurn true, newItem` is simpler.

File: server/handler.go [hunk 18]
```diff
@@ -111,14 +111,24 @@ func (h *Handler) GetHotReadRegions() *core.StoreHotRegionInfos {
 	return c.getHotReadRegions()
 }
 
-// GetHotWriteStores gets all hot write stores status
-func (h *Handler) GetHotWriteStores() map[uint64]uint64 {
-	return h.s.cluster.cachedCluster.getStoresWriteStat()
+// GetHotBytesWriteStores gets all hot write stores status
```
Comment (disksing) [inline server/handler.go]: _status_ -> _stats_ or _statistics_? BTW, sentences should end with a period.

File: server/core/store.go [hunk 19]
```diff
@@ -424,42 +429,122 @@ func (s *StoresInfo) SetRegionSize(storeID uint64, regionSize int64) {
...
+func newRollingStoreStats() *RollingStoreStats {
+	return &RollingStoreStats{
+		bytesWriteRate: NewRollingStats(storeStatsRollingWindows),
+		bytesReadRate:  NewRollingStats(storeStatsRollingWindows),
+		keysWriteRate:  NewRollingStats(storeStatsRollingWindows),
+		keysReadRate:   NewRollingStats(storeStatsRollingWindows),
+	}
+}
+
+// Observe records current statistics
```
Comment (disksing) [inline server/core/store.go]: Add a period at the end of sentence.

**Comment by disksing** [comment]
LGTM. PTAL @Connor1996

**Comment by nolouch** [comment]
PTAL @Connor1996

**Pull request merged**